docs/src/ref_impl/installer.md

```diff
@@ -5,67 +5,66 @@
-4. Write code for the installer in `ghaf.installer.installerCode`.
-
-```nix
-{config, ...}: {
-  ghaf.installer = {
-    enable = true;
-    imgModules = [
-      nixos-generators.nixosModules.raw-efi
-    ];
-    enabledModules = ["flushImage"];
-    installerCode = ''
-      echo "Starting flushing..."
-      if sudo dd if=${config.system.build.${config.formatAttr}} of=/dev/${config.ghaf.installer.installerModules.flushImage.providedVariables.deviceName} conv=sync bs=4K status=progress; then
-          sync
-          echo "Flushing finished successfully!"
-          echo "Now you can detach installation device and reboot to Ghaf."
-      else
-          echo "Some error occured during flushing process, exit code: $?."
-          exit
-      fi
-    '';
-  };
-}
-```
+Ghaf does not have conventional (imperative, non-declarative) installer, which
+you might expect in most of the conventional linux distros. Instead of this we
+have nixos-like non-interactive, declarative. In other words instead of clicking
+similar options each time you try to install system, you'll need to configure
+system once and deploy this configuraiton to desired machines. This done in a
+way that you need to use [ghaf as a library](./ghaf-based-project.md) in order to install system.
+
+This done via awesome tool
+[nixos-anywhere](https://github.com/nix-community/nixos-anywhere). There are two
+separate ways to use it: manually (by following 
+[their guide](https://nix-community.github.io/nixos-anywhere/quickstart.html)),
+or by using ghaf installation wizard.
+
+If you aware of NixOS installation process then this approach
+will seem familiar. Basically they are very similar, except system configuration
+is done in the familiar environment.
+
+## Via wizard
+
+Ghaf wizard will guide you through the process of creation of system
+configuration that you need.
+
+## Manual
+
+For now manual way is the only way to install ghaf, as wizard currently under
+development.
+
+To install ghaf manuall you need to:
 
-After that you can build an installer image using this command:
+1. Create your own flake via ghaf template
 
 ```sh
-nix build .#nixosConfigurations.<CONFIGURATION>.config.system.build.installer
+nix flake init -t github:tiiuae/ghaf#target-x86_64-generic
 ```
 
-## Adding Installer Modules
+2. Edit it according to your preferences.
 
-To add an installer module, replace the corresponding placeholders with your code and add this to your configuraiton:
+3. Set value of `ghaf.installer.sshKeys` in order to get installer image.
 
-```nix
-ghaf.installer.installerModules.<MODULE_NAME> = {
-  requestCode = ''
-    # Your request code written in Bash
-  '';
-  providedVariables = {
-    # Notice the dollar sign before the actual variable name in Bash.
-    <VARIABLE_NAME> = "$<VARIABLE_NAME>";
-  };
-};
+4. Build installer image.
+
+```sh
+nix build .#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.installer
```

Comment:
There's an error here: 

```
$ nix build .#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.installer
error: flake 'git+file:///home/vilvo/installer-PR340' does not provide attribute 'packages.x86_64-linux.nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.installer', 'legacyPackages.x86_64-linux.nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.installer' or 'nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.installer'
$ nix build .#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.binsh
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.bootStage1
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.bootStage2
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.earlyMountScript
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.etc
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.etcActivationCommands
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.extraUtils
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.fileSystems
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.initialRamdisk
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.initialRamdiskSecretAppender
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.installBootLoader
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.kernel
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.manual
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.nixos-enter
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.nixos-generate-config
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.nixos-install
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.nixos-option
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.nixos-rebuild
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.raw
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.setEnvironment
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.toplevel
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.units
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.vm
.#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.vmWithBootLoader
```

So it indeed is missing the attribute `.installer`

Reply:
I realized that the target is still in the PR branch but even that is missing the `.installer` target - I guessed `.installerTest` next but also that fails:
```
[vilvo@carrie:~/installer-PR340]$ nix flake init -t github:remimimimimi/ghaf/new-installation-approach#target-x86_64-generic
[vilvo@carrie:~/installer-PR340]$ nix build .#nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.installTest
warning: Git tree '/home/vilvo/installer-PR340' is dirty
error: The option `fileSystems."/".device' has conflicting definition values:
       - In `<unknown-file>': "/dev/pool/root"
       - In `/nix/store/3b1m8wjf30nh03viwpcihs56yrz07jqv-source/formats/raw.nix': "/dev/disk/by-label/nixos"
       Use `lib.mkForce value` or `lib.mkDefault value` to change the priority on any of these definitions.
(use '--show-trace' to show detailed location information
```

Reply:
That's probably because you need to change URL for Ghaf input in the initialized flake to path to repo with this branch. @vilvo

Reply:
Right. That helped a bit further. Thanks. 

This part (or broken template) makes the final step - here https://github.com/tiiuae/ghaf/pull/340/files#diff-da86839ca7970b827809fcbac9339d84ced530dece6f31bed56d235ff81c81d0R69 - break.
Can you identify what's causing the `error: attribute 'formatAttr' missing`?

```
[vilvo@carrie:~/installer-pr340]$ nix flake show
warning: creating lock file '/home/vilvo/installer-pr340/flake.lock'
warning: Git tree '/home/vilvo/installer-pr340' is dirty
git+file:///home/vilvo/installer-pr340
├───formatter
│   └───x86_64-linux: package 'alejandra-3.0.0'
├───nixosConfigurations
│   └───PROJ_NAME-ghaf-debug: NixOS configuration
└───packages
    └───x86_64-linux
error: attribute 'formatAttr' missing

       at /nix/store/qm6jvvjwm49zhqjdcv6gnj3p2gbbnf43-source/flake.nix:85:122:

           84|         };
           85|         packages.x86_64-linux.PROJ_NAME-ghaf-debug = self.nixosConfigurations.PROJ_NAME-ghaf-debug.config.system.build.${self.nixosConfigurations.PROJ_NAME-ghaf-debug.config.formatAttr};
             |                                                                                                                          ^
           86|       }
(use '--show-trace' to show detailed location information)

[vilvo@carrie:~/installer-pr340]$ ssh 172.20.10.4
Last login: Tue Nov 14 08:02:07 2023 from 172.20.10.3

[root@ghaf-host:~]#
logout
Connection to 172.20.10.4 closed.

[vilvo@carrie:~/installer-pr340]$ nix run github:numtide/nixos-anywhere -- --flake .#nixosConfigurations.PROJ_NAME-ghaf-debug root@172.20.10.4
warning: Git tree '/home/vilvo/installer-pr340' is dirty
error: flake 'git+file:///home/vilvo/installer-pr340' does not provide attribute 'packages.x86_64-linux.nixosConfigurations."nixosConfigurations.PROJ_NAME-ghaf-debug".config.system.build.diskoScript', 'legacyPackages.x86_64-linux.nixosConfigurations."nixosConfigurations.PROJ_NAME-ghaf-debug".config.system.build.diskoScript' or 'nixosConfigurations."nixosConfigurations.PROJ_NAME-ghaf-debug".config.system.build.diskoScript'

[vilvo@carrie:~/installer-pr340]$ nix run github:numtide/nixos-anywhere -- --flake .#packages.x86_64-linux.PROJ_NAME-ghaf-debug root@172.20.10.4
warning: Git tree '/home/vilvo/installer-pr340' is dirty
error: flake 'git+file:///home/vilvo/installer-pr340' does not provide attribute 'packages.x86_64-linux.nixosConfigurations."packages.x86_64-linux.PROJ_NAME-ghaf-debug".config.system.build.diskoScript', 'legacyPackages.x86_64-linux.nixosConfigurations."packages.x86_64-linux.PROJ_NAME-ghaf-debug".config.system.build.diskoScript' or 'nixosConfigurations."packages.x86_64-linux.PROJ_NAME-ghaf-debug".config.system.build.diskoScript'
```

My trial of adapting the instructions is available here https://github.com/vilvo/custom-ghaf/commit/05010604553137550656174428639b3a992aaddf

Reply:
You're trying to pass a package to nixos-anywhere, but it accepts as input name of the configuration, e.g you shall write `PROJ_NAME-ghaf-debug` instead of `packages.x86_64-linux.PROJ_NAME-ghaf-debug`.

And regarding the error, this happens because I forgot to change template in the same way as I done in [targets](https://github.com/tiiuae/ghaf/pull/340/files#diff-3aeeebee855bcdaed23bda0d9c30f96541d8a4dae4c681431213407282025890
).